### PR TITLE
Implement iOS one-click deploy.

### PR DESCRIPTION
**Pull request by bruvzg**
Updated and improved version of https://github.com/godotengine/godot/pull/33086

Implements https://github.com/godotengine/godot-proposals/issues/3506, one click deploy and remote debug for iOS devices and simulator.

Screenshot 2022-12-28 at 09 08 50

To use:
- Full `Xcode` (not command line tools) should be installed.
- `ios-deploy` should be installed and path added to the editor settings (`Export ⇾ iOS ⇾ iOSDeploy`).
  - Simulator should be running.
  - Device should be in [developer mode](https://developer.apple.com/documentation/xcode/enabling-developer-mode-on-a-device), have developer account added to trusted on a device (`Settings ⇾ General ⇾ VPN & Device Management`), and screen unlocked.
  - For remote debugging on a real device, make sure the device is on the same network and a correct network interface is selected in the editor settings (`Network ⇾ Debug ⇾ Remote Host`). By default, the editor is listening for localhost connections only.

Note: official export templates do not include arm64 simulator binary due to OSXCross limitations, so to use simulator on Apple Silicon Macs a custom build should be used (build flags `platform=ios ios_simulator=yes arch=arm64`).

### CODE REVIEW

File: editor/export/editor_export_platform.h [hunk 1]
```diff
@@ -220,6 +220,8 @@ class EditorExportPlatform : public RefCounted {
 		DEBUG_FLAG_REMOTE_DEBUG_LOCALHOST = 4,
 		DEBUG_FLAG_VIEW_COLLISIONS = 8,
 		DEBUG_FLAG_VIEW_NAVIGATION = 16,
+		DEBUG_FLAG_IOS_NO_ARCHIVE = 32,
+		DEBUG_FLAG_IOS_SIMULATOR = 64,
```
Comment (akien-mga) [inline editor/export/editor_export_platform.h:224]: Would there be away to implement this without having iOS specific values in the generic interface?
E.g. adding some kind of threshold value for platform-specific flags which can be defined in each platform implementation? Like `DEBUG_FLAGS_PLATFORM_SPECIFIC_THRESHOLD = 16` and then in iOS code they're defined as `1 << (DEBUG_FLAGS_PLATFORM_SPECIFIC_THRESHOLD + 1)` and `1 << (DEBUG_FLAGS_PLATFORM_SPECIFIC_THRESHOLD + 2)`.

Comment (bruvzg) [reply editor/export/editor_export_platform.h:224]: It is not necessary to have it in debug flags, so I guess I'll change it to use internal method with the extra args, similar to `export_project_helper` in Android export.

**Review by Senat2420 — APPROVED** [review]

**Comment by fire** [comment]
@punto- You mentioned this in a chat. Feel free to test it.

**Comment by gianmichelemariani** [comment]
Is this making it into official builds anytime soon? 4.1 ?

**Comment by Calinou** [comment]
> Is this making it into official builds anytime soon? 4.1 ?

4.1 is in feature freeze, so this won't be merged until after 4.1.

File: platform/ios/export/export.cpp [hunk 2]
```diff
@@ -38,6 +38,11 @@ void register_ios_exporter_types() {
 }
 
 void register_ios_exporter() {
+#ifdef MACOS_ENABLED
+	EDITOR_DEF("export/ios/iosdeploy", "");
+	EditorSettings::get_singleton()->add_property_hint(PropertyInfo(Variant::STRING, "export/ios/iosdeploy", PROPERTY_HINT_GLOBAL_FILE, "*"));
```
Comment (akien-mga) [inline platform/ios/export/export.cpp:43]: Might be more readable with underscore, and consistent with the dash in the upstream name: https://github.com/ios-control/ios-deploy

This should also ensure it's capitalized as "iOS Deploy" and not "Iosdeploy".
```suggestion
	EDITOR_DEF("export/ios/ios_deploy", "");
	EditorSettings::get_singleton()->add_property_hint(PropertyInfo(Variant::STRING, "export/ios/ios_deploy", PROPERTY_HINT_GLOBAL_FILE, "*"));
```

**Review by akien-mga — APPROVED** [review]
Looks good to me from a cursory check, though TIWAGOS, I'm not an expert.
Should be good to merge after 4.1 is released, so we can document it and test it throughly for 4.2.

**Comment by YuriSizov** [comment]
@bruvzg Needs a rebase and addressing Akien's comment.

**Comment by YuriSizov** [comment]
Thanks!

**Comment by shafnaz** [comment]
Nice!

**Comment by vanoak** [comment]
First, this works great! A huge timesaver! (Godot v4.2.stable - macOS 14.1.0 - GLES3 (Compatibility) - ANGLE (Apple, ANGLE Metal Renderer: Apple M1 Pro, Version 14.1 (Build 23B74)) - Apple M1 Pro (10 Threads))

The issue is that even in GLES3 (Compatibility) the "ALERT! Your GPU (Apple A9 GPU) does not support the following issues which are required to use Vulkan-based renderers in Godot" shows up which is unexpected because I thought that the GLES was not Vulkan based.

**Comment by Calinou** [comment]
> The issue is that even in GLES3 (Compatibility) the "ALERT! Your GPU (Apple A9 GPU) does not support the following issues which are required to use Vulkan-based renderers in Godot" shows up which is unexpected because I thought that GLES3 was not Vulkan based.

Make sure your `mobile` rendering method override is also changed to Compatibility: https://github.com/godotengine/godot/pull/72461

**Comment by vanoak** [comment]
Fixed. Thanks @Calinou.

**Comment by mellowminx** [comment]
@bruvzg Ok thank you for the note! I'm using Xcode 14.2, so I'll try it with another device using an older iOS version and report back :)

**Comment by mellowminx** [comment]
Following up on my previous comment https://github.com/godotengine/godot/pull/70662#issuecomment-1845958520 --

I tried with my older iPad, but still haven't been able to get one-click-deploy to iOS to work. The button for it is still grayed out and says "No Remote Debug export presets configured".

I'm using:
Godot 4.2
Xcode 14.2
MacOS 12.7.1 (Monterey)
Macbook Pro 16-inch M1 Pro, year 2021
**iPad Pro 12.9 inch Model A1652, year 2015, iOS 14.8**

I've set up the iOS export preset and set it to "Runnable"
My iPad is attached to my laptop using USB-C adaptor to USB-A to lightning cable
My iPad has "developer mode" enabled, screen is unlocked
I've reloaded the project, restarted Godot, unplugged and reconnected the USB-C cable
I've double-checked that my Macbook and iPad are both using the same wifi network

**Comment by Calinou** [comment]
> Are there more instructions on this using the simulator on Apple silicon? Perhaps this is the wrong post but I'm finding my builds in Xcode for simulators fail with this: Screenshot 2023-12-18 at 7 41 04 PM

Follow the instructions in https://github.com/godotengine/godot/issues/62929.

**Comment by ztc0611** [comment]
Information about this should really be added to [the one click deploy docs page](https://docs.godotengine.org/en/stable/tutorials/export/one-click_deploy.html).

**Comment by bruvzg** [comment]
> why exclude the simulator builds from the official template? 

OSXcross used to build official templates (on Linux) can't build it.

**Comment by leoyanis** [comment]
Quick question guys, why does the remote debug not work if i do the one click deploy on iOS? The debug console does not print anything and if i do changes in the scene it does on update on the phone, it does work on Android via cable though.

**Pull request merged**